data/src/history.rs

```diff
@@ -238,9 +396,30 @@ impl History {
             } => {
                 let server = server.clone();
                 let kind = kind.clone();
+                let unread_messages =
+                    message_reference.and_then(|message_reference| match message_reference {
+                        isupport::MessageReference::None => Some(messages.split_off(0)),
+                        _ => messages
+                            .iter()
+                            .rev()
+                            .position(|message| message_reference == *message)
+                            .map(|reference_position| {
+                                messages.split_off(messages.len() - reference_position)
+                            }),
+                    });
+                let opened_at = if let Some(ref unread_messages) = unread_messages {
+                    unread_messages.iter().fold(
+                        Posix::now(),
+                        |earliest_received_at, unread_message| {
+                            std::cmp::min(earliest_received_at, unread_message.received_at)
+                        },
+                    )
+                } else {
+                    Posix::now()
+                };
                 let messages = std::mem::take(messages);
 
-                *self = Self::partial(server.clone(), kind.clone(), Posix::now());
+                *self = Self::partial(server.clone(), kind.clone(), unread_messages, opened_at);
```

Comment:
I'm not sure I follow what `unread_messages` and `opened_at` means here when we convert from a full "opened" buffer history to a partial "closed" buffer history. 

Why are we calling `make_partial` from a chat history event? The idea behind our history API and full vs partial is that these are full / partial based on if the underlying buffer is open / closed.

- Partial exists as a way to store messages received on a closed buffer which are then fully flushed to disk.
- Full for open buffers has all messages in it and we append new messages. When a buffer is closed, we flush this entire history to disk.

This lifecycle is managed by `Manager::track` which is called after processing any `Dashboard` messages, where we see which set of buffers are open.

Shouldn't chathistory messages just feed in to the existing full / partial history that exists for a buffer? I don't think chathistory should break this lifecycle.

Reply:
This is mainly a continuation of the hack above, and is the second thing I was trying to highlight needed review.  I don't disagree with anything you're saying here, crimes are being committed which need to be rectified.  But let me lay out the functionality that the crimes provide because I'm not yet sure how else to get at it.

When joining a channel after launch, we want to get all messages since we last closed Halloy.  To do that properly, we need to know what the last message we received for that channel was, but we can't do that with having loaded the message history for the channel.  We can request the `LATEST` however many messages, but there will often be duplicates in there we don't really want to process and we won't know if they're duplicates until we can compare them to the message history.  That messes with our ability to tell the user whether the channel has any new messages or not, and I wasn't able to see any way around it.  The process of loading a channel would also become more complicated in that scenario (deduplication would need to be done on channel load, along with any additional requests needed to fill in chat history that wasn't covered by the initial request), but that part is doable.

So, the hack that's being performed now is to load the history for every channel joined, use the loaded history to form the chathistory request for the latest messages, position received messages, and deduplicate, then make the loaded history partial.  All this junk right here is a way to convert a loaded history with all the chathistory request messages added into a proper partial history that we'd see if the chathistory messages had been loaded into a proper, fresh partial history.